### PR TITLE
Support Custom span-id and trace-id generator

### LINKED ISSUE #677: Custom Trace and Span IDs Generation
Custom IDs would allow users to set IDs that they need right at the beginning of a span. This is important if the downstream services or eventual backend that receives these spans need them to be in a certain format.

As per the specs:
https://github.com/open-telemetry/opentelemetry-specification/blob/b67faa4e0106cac194a837e3d419ae4fab9fc1c4/specification/trace/sdk.md#id-generators
```The SDK MUST provide a mechanism for customizing the way IDs are generated for both the TraceId and the SpanId.```

This is how other languages support it:
JS: https://github.com/open-telemetry/opentelemetry-js/blob/b7d6e74c5bec6a24f5d617d31402ea95c54f126f/packages/opentelemetry-core/src/trace/IdGenerator.ts

Java: https://github.com/open-telemetry/opentelemetry-java/blob/485cc52c6ff1bb3fa12cd950d6eb97d3ceae6e41/sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/IdsGenerator.java

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #677 

## Changes
  - Add abstract interface for a span-id and trace-id generation.
  - Convert existing logic for random span-id and trace-id as the default implementation for the above interface.
   - Modify TracerProvider and TracerContext constructor to support injecting the custom id generator.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: sdk/test/trace/tracer_test.cc [hunk 1]
```diff
@@ -45,6 +45,24 @@ class MockSampler final : public Sampler
   nostd::string_view GetDescription() const noexcept override { return "MockSampler"; }
 };
 
+/**
+ * A Mock Custom Id Generator
+ */
+class MockIdGenerator : public IdGenerator
```
Comment (eyjohn) [inline sdk/test/trace/tracer_test.cc:51]: nit: Google Mock would be great to use here, I noticed we don't have it in the repo but EXPECT_CALL and return values can all be done in the test body which would be nice

File: sdk/test/trace/tracer_test.cc [hunk 2]
```diff
@@ -132,6 +152,30 @@ TEST(Tracer, StartSpanSampleOff)
...
+  char rec_trace[32], exp_trace[32];
+  char rec_span[16], exp_span[16];
+
+  cur_span_data->GetTraceId().ToLowerBase16(rec_trace);
+  id_generator->GenerateTraceId().ToLowerBase16(exp_trace);
+
+  cur_span_data->GetSpanId().ToLowerBase16(rec_span);
+  id_generator->GenerateSpanId().ToLowerBase16(exp_span);
+
+  ASSERT_EQ(std::string(rec_trace, sizeof(rec_trace)), std::string(exp_trace, sizeof(exp_trace)));
```
Comment (eyjohn) [inline sdk/test/trace/tracer_test.cc:175]: nit: EXPECT_EQ

File: sdk/include/opentelemetry/sdk/trace/id_generator.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+namespace trace
+{
+
+/** IdGenerator provides an interface for generating Trace Id and Span Id */
+class IdGenerator
+{
+
+public:
+  /** Returns a SpanId represented by opaque 128-bit trace identifier */
+  virtual opentelemetry::trace::SpanId GenerateSpanId() noexcept = 0;
```
Comment (eyjohn) [inline sdk/include/opentelemetry/sdk/trace/id_generator.h:32]: Should the contract enforce constness or not?

I can foresee an implementation that does mutate internal state, a simple implementation being an incremental counter that generates an ID based on the counter. But I'm not sure whether this is by design or not?

It seems Sampler likewise doesn't use `const`

File: sdk/include/opentelemetry/sdk/trace/id_generator.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  /** Returns a TraceId represented by opaque 64-bit trace identifier */
+  virtual opentelemetry::trace::TraceId GenerateTraceId() noexcept = 0;
+};
+
+class RandomIdGenerator : public IdGenerator
```
Comment (eyjohn) [inline sdk/include/opentelemetry/sdk/trace/id_generator.h:38]: nit: should the derived class live in a separate file? it does for samplers

**Review by eyjohn — APPROVED** [review]
Overall looks great to land, some minor nitpicks,

Minor question about constness

File: sdk/test/trace/tracer_test.cc [hunk 5]
```diff
@@ -132,6 +152,30 @@ TEST(Tracer, StartSpanSampleOff)
...
+
+  tracer->StartSpan("span 1")->End();
+  auto spans          = span_data->GetSpans();
+  auto &cur_span_data = spans.at(0);
+
+  char rec_trace[32], exp_trace[32];
+  char rec_span[16], exp_span[16];
+
+  cur_span_data->GetTraceId().ToLowerBase16(rec_trace);
+  id_generator->GenerateTraceId().ToLowerBase16(exp_trace);
```
Comment (pyohannes) [inline sdk/test/trace/tracer_test.cc:170]: I think we can compare ids directly? Like `EXPECT_EQ(cur_span_data->GetTraceId(), id_generator->GenerateTraceId());`?

File: sdk/include/opentelemetry/sdk/trace/id_generator.h (hunk 3 above)
Comment (pyohannes) [reply sdk/include/opentelemetry/sdk/trace/id_generator.h:32]: I agree. We shouldn't use `const`, as we can't assume the `IdGenerator` to be stateless.

File: sdk/include/opentelemetry/sdk/trace/tracer_provider.h [hunk 6]
```diff
@@ -27,11 +27,14 @@ class TracerProvider final : public opentelemetry::trace::TracerProvider
...
-                              opentelemetry::sdk::resource::Resource::Create({}),
-                          std::unique_ptr<Sampler> sampler =
-                              std::unique_ptr<AlwaysOnSampler>(new AlwaysOnSampler)) noexcept;
+  explicit TracerProvider(
+      std::unique_ptr<SpanProcessor> processor,
+      opentelemetry::sdk::resource::Resource resource =
+          opentelemetry::sdk::resource::Resource::Create({}),
+      std::unique_ptr<Sampler> sampler = std::unique_ptr<AlwaysOnSampler>(new AlwaysOnSampler),
+      std::unique_ptr<opentelemetry::sdk::trace::IdGenerator> id_generator =
+          std::unique_ptr<opentelemetry::sdk::trace::IdGenerator>(
```
Comment (ThomsonTan) [inline sdk/include/opentelemetry/sdk/trace/tracer_provider.h:36]: Is it better to make `IdGenerator` a global static? Is there any scenario to assign a `TraceProvider` a dedicated `IdGenerator`?

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/trace/tracer_provider.h:36]: As per the [specs](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#tracer-creation), `IdGenerator` should be managed by TracerProvider
``
Configuration (i.e., SpanProcessors, IdGenerator, SpanLimits and Sampler) MUST be managed solely by the TracerProvider and it MUST provide some way to configure all of them that are implemented in the SDK, at least when creating or initializing it.
``

File: sdk/test/trace/tracer_test.cc (hunk 5 above)
Comment (lalitb) [reply sdk/test/trace/tracer_test.cc:170]: Good point. It is bitwise comparison of `TraceId` data members ( array of uint8_t ) and it should work.

File: sdk/include/opentelemetry/sdk/trace/id_generator.h (hunk 3 above)
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/trace/id_generator.h:32]: We shouldn't enforce constness, which is how `IdGenerator` is currently implemented. Let me know if you foresee any further changes?

File: sdk/include/opentelemetry/sdk/trace/id_generator.h (hunk 4 above)
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/trace/id_generator.h]: Good point, moved derived class to separate header.

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**